### ISSUE TITLE
Refactor: Separate stateful fit and transform lifecycles in ArnioCleaner

### ISSUE THREAD

**Issue by naina-bhatnagar** [im-anishraj/arnio#1272]
### Current situation

In `arnio/integrations/sklearn.py`, the `ArnioCleaner.fit()` method only validates column layouts and stores feature metadata. The actual execution engine is invoked dynamically inside `transform()` via:

> ```python
> cleaned_ar_frame = run_pipeline(ar_frame, self.steps)
> ```

Because the execution sequence evaluates parameters from scratch inside `transform()`, any pipeline cleaning steps that compute statistical properties (like group means for imputation, variance calculations, or value-to-index mappings) re-estimate their data metrics completely independently on whatever dataset split is passed into `transform()`.

### Proposed change

I propose restructuring the execution lifecycle inside the wrapper to separate state estimation from state application:

1. **State Gathering:** Update `fit()` to inspect the training dataframe using an Arnio frame pass, capturing a frozen dictionary of learned statistical variables:
> ```python
> self.learned_states_ = {} # Populated by C++ backend analysis
> ```

2. **State Mapping:** Update `transform()` to pass this frozen dictionary directly down into the pipeline's execution logic to guarantee stable calculations:
> ```python
> cleaned_ar_frame = run_pipeline(ar_frame, self.steps, state=self.learned_states_)
> ```

### Expected benefit

- **Prevents Data Leakage:** Ensures that preprocessing constraints are learned strictly from the training dataset and strictly applied to validation or test data configurations.
- **Enables Production Inference:** Guarantees that single-row prediction streams do not crash or corrupt values when running steps that rely on running statistics (such as mean imputation or Z-score outlier filtering).
- **Architecture Compliance:** Aligns `ArnioCleaner` directly with standard `scikit-learn` lifecycle requirements for transformers.

### Risk assessment

- [x] This change is backwards-compatible.
- [x] This change does not alter the public API.

### Additional context

I would love to work on this issu as a part of GSSoC'26 ! Could you please assign it to me?

**Comment by nancy-verma780** [comment]
Hi @naina-bhatnagar, I would love to work on this issue! The proposal to separate state estimation in `fit()` from state application in `transform()` makes total sense and prevents critical data leakage. 

Could you please assign this issue to me? I have a solution ready to submit. Thanks!

**Comment by im-anishraj** [comment]
Thanks for the proposal. I’m leaving this as needs-maintainer rather than assigning it right now because it is a larger sklearn lifecycle/API design change.

Before a PR, we need to define which Arnio steps are actually stateful today, what learned state format they expose, how that state is serialized/applied, and how this interacts with the existing `ArnioCleaner` contract. Please don’t open a PR until that design is accepted; this likely needs a focused design discussion first, not an immediate implementation.

**Comment by nancy-verma780** [comment]
Hi @im-anishraj and @naina-bhatnagar,

I completely agree with the need to define the design contract before opening a PR. Here is a proposed architecture for handling stateful pipeline steps within the `ArnioCleaner` scikit-learn lifecycle:

### 1. State Format & Serialization Contract
The `self.learned_states_` can be structured as a nested dictionary where keys are individual step identifiers (or execution indices) and values represent their specific learned properties:
```python
self.learned_states_ = {
    "step_0_impute": {"column_means": {"age": 30.5, "salary": 55000.0}},
    "step_1_scale": {"z_score_params": {"age": {"mean": 30.5, "std": 5.2}}}
}
```

### 2. Identifying Stateful Steps
Currently, stateful steps include primitive transforms that rely on dataset-wide metrics (e.g., mean/median imputation, value-to-index mappings for categorical encoders, and Z-score outlier boundaries). Purely stateless operations (like `cast_types` or drop operations) will simply yield an empty schema state `{}`.

### 3. Execution Integration with `run_pipeline`
* **During `fit()`**: We invoke an analysis pass over the `ArnioFrame`. For any step that inherits from a stateful base class, we capture its generated parameters into `self.learned_states_`.
* **During `transform()`**: We pass the frozen `self.learned_states_` dict down to `run_pipeline(ar_frame, self.steps, state=self.learned_states_)`. The backend logic will prioritize applying these frozen parameters over calculating metrics dynamically from the validation slice.

I would love to spearhead this refactor and establish these lifecycle hooks. Could you please assign this issue to me so I can finalize this layout in a PR?

**Comment by naina-bhatnagar** [comment]
Hi @im-anishraj, after doing a line-by-line review of `arnio/cleaning.py`, I want to provide an update on this proposal. I actually don't think we need an architectural refactor for state tracking at this stage.



### The Reason

If we analyze the library's current cleaning steps, they are completely stateless:

1. **Scalar Operations:** Steps like `fill_nulls` only accept fixed static scalars (e.g., `value=0`), meaning they don't calculate data-dependent properties.

2. **Row/Element Operations:** Steps like `strip_whitespace`, `cast_types`, and `normalize_case` operate row-by-row and require zero dataset-wide context.

3. **Outlier Handling:** While `winsorize_outliers` computes quantiles, it currently runs entirely as a Pandas fallback and doesn't hook into the native C++ pipeline execution array (`run_pipeline`). 



### Conclusion

Since the current underlying engine does not support stateful components or dynamic strategies (like `value="mean"`), there is no actual data leakage happening inside our native execution tracks. 



Therefore, I am closing this issue.

**Comment by im-anishraj** [comment]
Thanks for the follow-up. I agree with the conclusion: current built-in cleaning steps do not have a fit-time learned-state contract, so this larger lifecycle refactor is not actionable right now.

I’m closing this as not planned for the current API. We can revisit with a narrower issue if/when Arnio adds stateful steps such as learned imputers, encoders, or scalers.
